File: common/src/main/java/me/kvalbrus/multibans/common/punishment/MultiPunishment.java
package me.kvalbrus.multibans.common.punishment;

import java.util.Arrays;
import java.util.List;
import me.kvalbrus.multibans.api.DataProvider;
import me.kvalbrus.multibans.api.punishment.Punishment;
import me.kvalbrus.multibans.api.punishment.creator.OnlinePunishmentCreator;
import me.kvalbrus.multibans.api.punishment.creator.PunishmentCreator;
import me.kvalbrus.multibans.api.punishment.target.OnlinePunishmentTarget;
import me.kvalbrus.multibans.api.punishment.target.PunishmentTarget;
import me.kvalbrus.multibans.api.punishment.PunishmentType;
import me.kvalbrus.multibans.api.punishment.TemporaryPunishment;
import me.kvalbrus.multibans.common.managers.MultiBansPluginManager;
import me.kvalbrus.multibans.common.managers.PluginManager;
import me.kvalbrus.multibans.common.permissions.Permission;
import me.kvalbrus.multibans.common.punishment.punishments.MultiPermanentlyBan;
import me.kvalbrus.multibans.common.punishment.punishments.MultiPermanentlyBanIp;
import me.kvalbrus.multibans.common.punishment.punishments.MultiPermanentlyChatMute;
import me.kvalbrus.multibans.common.punishment.punishments.MultiTemporaryBan;
import me.kvalbrus.multibans.common.punishment.punishments.MultiTemporaryBanIp;
import me.kvalbrus.multibans.common.punishment.punishments.MultiTemporaryChatMute;
import me.kvalbrus.multibans.common.utils.ReplacedString;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class MultiPunishment implements Punishment {

    private final PluginManager pluginManager;

    private final PunishmentType type;

    private final String id;

    private final PunishmentTarget target;

    private final PunishmentCreator creator;

    private final long createdDate;

    private String reason;

    private String comment;

    private List<String> servers;

    public MultiPunishment(@NotNull PluginManager pluginManager,
                           @NotNull final PunishmentType type,
                           @NotNull final String id,
                           @NotNull final PunishmentTarget target,
                           @NotNull final PunishmentCreator creator,
                           final long createdDate,
                           @Nullable String reason,
                           @Nullable String comment,
                           @NotNull List<String> servers) {
        this.pluginManager = pluginManager;
        this.type = type;
        this.id = id;

        this.target = target;
        this.target.setPunishment(this);

        this.creator = creator;
        this.creator.setPunishment(this);

        this.createdDate = createdDate;
        this.reason = reason;
        this.comment = comment;
        this.servers = servers;
    }

    @NotNull
    public static <T extends Punishment> T constructPunishment(
        @NotNull PluginManager pluginManager,
        @NotNull PunishmentType type,
        @NotNull String id,
        @NotNull PunishmentTarget target,
        @NotNull PunishmentCreator creator,
        long createdDate,
        long startedDate,
        long duration,
        @Nullable String createdReason,
        @Nullable String comment,
        @Nullable PunishmentCreator cancellationCreator,
        long cancellationDate,
        @Nullable String cancellationReason,
        @NotNull List<String> servers,
        boolean cancelled) {
        Punishment punishment;
        switch (type) {
            case BAN:
                punishment = new MultiPermanentlyBan(pluginManager, id, target, creator, createdDate,
                    createdReason, comment, cancellationCreator, cancellationDate,
                    cancellationReason, servers, cancelled);
                break;

            case TEMP_BAN:
                punishment = new MultiTemporaryBan(pluginManager, id, target, creator, createdDate,
                    startedDate, duration, createdReason, comment,
                    cancellationCreator, cancellationDate, cancellationReason, servers, cancelled);
                break;

            case BAN_IP:
                punishment = new MultiPermanentlyBanIp(pluginManager, id, target, creator,
                    createdDate, createdReason, comment, cancellationCreator, cancellationDate,
                    cancellationReason, servers, cancelled);
                break;

            case TEMP_BAN_IP:
                punishment = new MultiTemporaryBanIp(pluginManager, id, target, creator, createdDate,
                    startedDate, duration, createdReason, comment, cancellationCreator,
                    cancellationDate, cancellationReason, servers, cancelled);
                break;

            case MUTE:
                punishment = new MultiPermanentlyChatMute(pluginManager, id, target, creator,
                    createdDate, createdReason, comment, cancellationCreator, cancellationDate,
                    cancellationReason, servers, cancelled);
                break;

            case TEMP_MUTE:
                punishment = new MultiTemporaryChatMute(pluginManager, id, target, creator,
                    createdDate, startedDate, duration, createdReason, comment, cancellationCreator,
                    cancellationDate, cancellationReason, servers, cancelled);
                break;

            default:
                throw new IllegalArgumentException("Punishment type hasn't its constructor");
        }

        return (T) punishment;
    }

    @Override
    public synchronized void activate() {
        this.pluginManager.activatePunishment(this);
        this.updateData();
        this.sendMessageAboutActivate();
    }

    @Override
    public synchronized void delete() {
        this.pluginManager.getDataProvider().deletePunishment(this);
        this.deleteData();
        this.sendMessageAboutDelete();
    }

    @NotNull
    public final PluginManager getPluginManager() {
        return this.pluginManager;
    }

    @NotNull
    @Override
    public final PunishmentType getType() {
        return this.type;
    }

    @NotNull
    @Override
    public final String getId() {
        return this.id;
    }

    @Override
    public final PunishmentTarget getTarget() {
        return this.target;
    }

    @Override
    public final PunishmentCreator getCreator() {
        return this.creator;
    }

    @Override
    public final long getCreatedDate() {
        return this.createdDate;
    }

    @Nullable
    @Override
    public final String getCreatedReason() {
        return this.reason;
    }

    @Nullable
    @Override
    public final String getComment() {
        return this.comment;
    }

    @NotNull
    @Override
    public final List<String> getServers() {
        return this.servers;
    }

    @Override
    public final synchronized void setCreatedReason(@Nullable String reason) {
        this.reason = reason;
        this.updateData();
        this.sendMessageAboutReasonChange();
    }

    @Override
    public final synchronized void setComment(@Nullable String comment) {
        this.comment = comment;
        this.updateData();
        this.sendMessageAboutCommentChange();
    }

    @Override
    public synchronized void setServers(@NotNull List<String> servers) {
        this.servers = servers;
        this.updateData();
    }

    @NotNull
    public abstract String getActivateMessageForListener();

    @NotNull
    public abstract String getActivateMessageForExecutor();

    @Nullable
    public abstract String getActivateMessageForTarget();

    @NotNull
    public abstract String getDeleteMessageForListener();

    @NotNull
    public abstract String getDeleteMessageForExecutor();

    @Nullable
    public abstract String getDeleteMessageForTarget();

    @NotNull
    public abstract String getReasonChangeMessageForExecutor();

    @NotNull
    public abstract String getReasonChangeMessageForListener();

    @NotNull
    public abstract String getCommentChangeMessageForExecutor();

    @NotNull
    public abstract String getCommentChangeMessageForListener();

    @NotNull
    public abstract Permission getPermissionForListener();

    @Override
    public int compareTo(@NotNull Punishment punishment2) {
        Punishment punishment1 = this;

        if (punishment1.getType() != punishment2.getType()) {
            return punishment1.getType().compareTo(punishment2.getType());
        }

        if (!punishment1.getTarget().getUniqueId().equals(punishment2.getTarget().getUniqueId())) {
            return punishment1.getTarget().getUniqueId().compareTo(punishment2.getTarget().getUniqueId());
        }

        if (punishment1 instanceof TemporaryPunishment t1 &&
            punishment2 instanceof TemporaryPunishment t2) {
            if (t1.getStartedDate() != t2.getStartedDate()) {
                return Long.compare(t1.getStartedDate(), t2.getStartedDate());
            }

            if (t1.getDuration() != t2.getDuration()) {
                return Long.compare(t1.getDuration(), t2.getDuration());
            }
        }

        if (!punishment1.getCreator().getName().equals(punishment2.getCreator().getName())) {
            return punishment1.getCreator().getName().compareTo(punishment2.getCreator().getName());
        }

        if (punishment1.getCreatedDate() != punishment2.getCreatedDate()) {
            return Long.compare(punishment1.getCreatedDate(), punishment2.getCreatedDate());
        }

        return punishment1.getId().compareTo(punishment2.getId());
    }

    public synchronized void updateData() {
        DataProvider dataProvider = this.getPluginManager().getDataProvider();
        if (dataProvider != null) {
            if (dataProvider.hasPunishment(this.getId())) {
                dataProvider.updatePunishment(this);
            } else {
                dataProvider.createPunishment(this);
            }
        }
    }

    public synchronized void deleteData() {
        DataProvider dataProvider = this.getPluginManager().getDataProvider();
        if (dataProvider != null && dataProvider.hasPunishment(this.getId())) {
            dataProvider.deletePunishment(this);
        }
    }

    private void sendMessageAboutActivate() {
        this.sendMessageToListeners(this.getActivateMessageForListener());
        this.sendMessageToCreator(this.getActivateMessageForExecutor());
        this.sendMessageToTarget(this.getActivateMessageForTarget());
    }

    private void sendMessageAboutDelete() {
        this.sendMessageToListeners(this.getDeleteMessageForListener());
        this.sendMessageToCreator(this.getDeleteMessageForExecutor());
        this.sendMessageToTarget(this.getDeleteMessageForTarget());
    }

    private void sendMessageAboutCommentChange() {
        this.sendMessageToListeners(this.getCommentChangeMessageForListener());
        this.sendMessageToCreator(this.getCommentChangeMessageForExecutor());
    }

    private void sendMessageAboutReasonChange() {
        this.sendMessageToListeners(this.getReasonChangeMessageForListener());
        this.sendMessageToCreator(this.getReasonChangeMessageForExecutor());
    }

    public final void sendMessageToListeners(String message) {
        ReplacedString listenMessage = new ReplacedString(message).replacePunishment(this);

        // Listener #1 - online players
        Arrays.stream(this.getPluginManager().getOnlinePlayers())
            .filter(player -> player.hasPermission(this.getPermissionForListener().getName()))
            .forEach(player -> player.sendMessage(listenMessage.string()));

        // Listener #2 - console
        if (this.getPluginManager() instanceof MultiBansPluginManager multiBansPluginManager) {
            if (multiBansPluginManager.getSettings().isConsoleLog()) {
                this.getPluginManager().getConsole().sendMessage(listenMessage.string());
            }
        }
    }

    public final void sendMessageToCreator(String message) {
        if (this.creator instanceof OnlinePunishmentCreator creator) {
            ReplacedString creatorMessage = new ReplacedString(message).replacePunishment(this);
            creator.sendMessage(creatorMessage.string());
        }
    }

    public final void sendMessageToTarget(String message) {
        if (this.target instanceof OnlinePunishmentTarget target) {
            ReplacedString targetMessage = new ReplacedString(message).replacePunishment(this);
            target.sendMessage(targetMessage.string());
        }
    }
}

File: common/src/main/kotlin/me/kvalbrus/multibans/common/punishment/MultiPunishment.kt
package me.kvalbrus.multibans.common.punishment

import me.kvalbrus.multibans.api.OnlinePlayer
import me.kvalbrus.multibans.api.punishment.Punishment
import me.kvalbrus.multibans.api.punishment.punishments.PunishmentType
import me.kvalbrus.multibans.api.punishment.TemporaryPunishment
import me.kvalbrus.multibans.api.punishment.action.Action
import me.kvalbrus.multibans.api.punishment.action.CreationAction
import me.kvalbrus.multibans.api.punishment.executor.OnlinePunishmentExecutor
import me.kvalbrus.multibans.api.punishment.executor.PunishmentExecutor
import me.kvalbrus.multibans.api.punishment.target.OnlinePunishmentTarget
import me.kvalbrus.multibans.api.punishment.target.PunishmentTarget
import me.kvalbrus.multibans.common.managers.MultiBansPluginManager
import me.kvalbrus.multibans.common.managers.PluginManager
import me.kvalbrus.multibans.common.permissions.Permission
import me.kvalbrus.multibans.common.punishment.action.MultiCreationAction
import me.kvalbrus.multibans.common.punishment.punishments.*
import me.kvalbrus.multibans.common.utils.Message
import me.kvalbrus.multibans.common.utils.ReplacedString
import java.util.*

abstract class MultiPunishment(
    pluginManager: PluginManager,
    type: PunishmentType,
    id: String,
    creationAction: CreationAction,
    comment: String,
    servers: List<String>) : Punishment {

    private var _comment: String
    private var _servers: List<String>

    protected val pluginManager: PluginManager

    final override val type: PunishmentType
    final override val id: String
    final override val creationAction: CreationAction

    init {
        this.pluginManager = pluginManager
        this.type = type
        this.id = id
        this.creationAction = creationAction
        this._comment = comment
        this._servers = servers

        this.target.punishment = this
        this.creator.punishment = this
    }

    final override val target: PunishmentTarget
        get() = this.creationAction.target

    final override val creator: PunishmentExecutor
        get() = this.creationAction.executor

    final override val createdDate: Long
        get() = this.creationAction.date

    final override var createdReason: String
        get() = this.creationAction.reason

        @Synchronized
        set(createdReason) {
            this.creationAction.reason = createdReason
            this.updateData()
            this.sendMessageAboutReasonChange()
        }

    final override var comment: String
        get() = this._comment

        @Synchronized
        set(comment) {
            this._comment = comment
            this.updateData()
            this.sendMessageAboutCommentChange()
        }

    final override var servers: List<String>
        get() = this._servers

        @Synchronized
        set(servers) {
            this._servers = servers
            this.updateData()
        }

    abstract val createMessageForListener: Message
    abstract val createMessageForExecutor: Message
    abstract val createMessageForTarget: Message
    abstract val deleteMessageForListener: Message
    abstract val deleteMessageForExecutor: Message
    abstract val deleteMessageForTarget: Message
    abstract val reasonChangeMessageForExecutor: Message
    abstract val reasonChangeMessageForListener: Message
    abstract val commentChangeMessageForExecutor: Message
    abstract val commentChangeMessageForListener: Message
    abstract val permissionForListener: Permission

    constructor(
        pluginManager: PluginManager,
        type: PunishmentType,
        id: String,
        target: PunishmentTarget,
        creator: PunishmentExecutor,
        createdDate: Long,
        createdReason: String,
        comment: String,
        servers: List<String>) :
            this(pluginManager, type, id,
                MultiCreationAction(id, 1, target, creator, createdDate, createdReason), comment, servers)

    @Synchronized
    override fun create() {
        this.updateData()
        this.pluginManager.createPunishment(this, this.creationAction)
        this.sendMessageAboutCreation()
    }

    @Synchronized
    override fun delete() {
        this.pluginManager.dataProvider?.deletePunishment(this)
        this.deleteData()
        this.pluginManager.deletePunishment(this)
        this.sendMessageAboutDelete()
    }

    override fun compareTo(other: Punishment?): Int {
        other?.let {
            val punishment1: Punishment = this
            if (punishment1.type != other.type) {
                return punishment1.type.compareTo(other.type)
            }

            if (punishment1.target.uniqueId != other.target.uniqueId) {
                return punishment1.target.uniqueId.compareTo(other.target.uniqueId)
            }

            if (punishment1 is TemporaryPunishment && other is TemporaryPunishment) {
                if (punishment1.startedDate != other.startedDate) {
                    return punishment1.startedDate.compareTo(other.startedDate)
                }

                if (punishment1.duration != other.duration) {
                    return punishment1.duration.compareTo(other.duration)
                }
            }

            if (punishment1.creator.name != other.creator.name) {
                return punishment1.creator.name.compareTo(other.creator.name)
            }

            if (punishment1.createdDate != other.createdDate) {
                return punishment1.createdDate.compareTo(other.createdDate)
            } else {
                return punishment1.id.compareTo(other.id)
            }
        }

        return -1
    }

    @Synchronized
    protected fun updateData() {
        val dataProvider = pluginManager.dataProvider

        if (dataProvider?.hasPunishment(id) == true) {
            dataProvider.savePunishment(this)
        } else {
            dataProvider?.createPunishment(this)
        }

    }

    @Synchronized
    protected fun deleteData() {
        val dataProvider = pluginManager.dataProvider
        if (dataProvider != null && dataProvider.hasPunishment(id)) {
            dataProvider.deletePunishment(this)
        }
    }

    protected fun sendMessageToListeners(message: Message) {
        val listenMessage = ReplacedString(message.getText()).replacePunishment(this)

        // Listener #1 - online players
        Arrays.stream(pluginManager.onlinePlayers)
            .filter { player: OnlinePlayer ->
                player.hasPermission(this.permissionForListener.perm) }
            .forEach { player: OnlinePlayer -> player.sendMessage(listenMessage.string()) }

        // Listener #2 - console
        if (pluginManager is MultiBansPluginManager) {
            if (pluginManager.settings.consoleLog) {
                pluginManager.getConsole().sendMessage(listenMessage.string())
            }
        }
    }

    protected fun sendMessageToCreator(message: Message) {
        if (this.creator is OnlinePunishmentExecutor) {
            val creatorMessage = ReplacedString(message.getText()).replacePunishment(this)
            (this.creator as OnlinePunishmentExecutor).sendMessage(creatorMessage.string())
        }
    }

    protected fun sendMessageToTarget(message: Message) {
        if (this.target is OnlinePunishmentTarget) {
            val targetMessage = ReplacedString(message.getText()).replacePunishment(this)
            (this.target as OnlinePunishmentTarget).sendMessage(targetMessage.string())
        }
    }

        private fun sendMessageAboutCreation() {
        sendMessageToListeners(this.createMessageForListener)
        sendMessageToCreator(this.createMessageForExecutor)
        sendMessageToTarget(this.createMessageForTarget)
    }

    private fun sendMessageAboutDelete() {
        sendMessageToListeners(this.deleteMessageForListener)
        sendMessageToCreator(this.deleteMessageForExecutor)
        sendMessageToTarget(this.deleteMessageForTarget)
    }

    private fun sendMessageAboutCommentChange() {
        sendMessageToListeners(this.commentChangeMessageForListener)
        sendMessageToCreator(this.commentChangeMessageForExecutor)
    }

    private fun sendMessageAboutReasonChange() {
        sendMessageToListeners(this.reasonChangeMessageForListener)
        sendMessageToCreator(this.reasonChangeMessageForExecutor)
    }

    companion object {
        fun <T : Punishment?> constructPunishment(
            pluginManager: PluginManager,
            type: PunishmentType,
            id: String,
            creationAction: CreationAction,
            activations: MutableList<Action> = mutableListOf(),
            deactivations: MutableList<Action> = mutableListOf(),
            startedDate: Long,
            duration: Long,
            comment: String,
            servers: List<String>,
            cancelled: Boolean): T {
            val punishment: Punishment = when (type) {
                PunishmentType.BAN -> MultiPermanentlyBan(
                    pluginManager, id, creationAction, activations, deactivations, comment, servers, cancelled)

                PunishmentType.TEMP_BAN -> MultiTemporaryBan(pluginManager, id, creationAction,
                    activations, deactivations, startedDate, duration, comment, servers, cancelled)

                PunishmentType.BAN_IP -> MultiPermanentlyBanIp(
                    pluginManager, id, creationAction, activations, deactivations, comment, servers, cancelled)

                PunishmentType.TEMP_BAN_IP -> MultiTemporaryBanIp(pluginManager, id, creationAction,
                    activations, deactivations, startedDate, duration, comment, servers, cancelled)

                PunishmentType.MUTE -> MultiPermanentlyChatMute(
                    pluginManager, id, creationAction, activations, deactivations, comment, servers, cancelled)

                PunishmentType.TEMP_MUTE -> MultiTemporaryChatMute(pluginManager, id, creationAction,
                    activations, deactivations, startedDate, duration, comment, servers, cancelled)

                PunishmentType.KICK -> Kick(pluginManager, id, creationAction, comment, servers)

                else -> throw IllegalArgumentException("Punishment type hasn't its constructor")
            }

            return punishment as T
        }
    }
}